src/main/java/com/google/firebase/database/connection/WebsocketConnection.java

```diff
@@ -338,67 +312,102 @@ public void onClose() {
         logger.debug("closed");
       }
       if (!isClosed) {
-        executorService.execute(new Runnable() {
-          @Override
-          public void run() {
-            onClosed();
-          }
-        });
+        // If the connection tear down was initiated by the higher-layer, isClosed will already
+        // be true. Nothing more to do in that case.
+        executorService.execute(
+            new Runnable() {
+              @Override
+              public void run() {
+                onClosed();
+              }
+            });
       }
     }
 
     @Override
-    public void onError(final WebSocketException e) {
+    public void onError(final Throwable e) {
+      if (e.getCause() != null && e.getCause() instanceof EOFException) {
+        logger.error("WebSocket reached EOF", e);
+      } else {
+        logger.error("WebSocket error", e);
+      }
       executorService.execute(
           new Runnable() {
             @Override
             public void run() {
-              if (e.getCause() != null && e.getCause() instanceof EOFException) {
-                logger.debug("WebSocket reached EOF.");
-              } else {
-                logger.debug("WebSocket error.", e);
-              }
               onClosed();
             }
           });
     }
+  }
 
-    @Override
-    public void onLogMessage(String msg) {
-      if (logger.logsDebug()) {
-        logger.debug("Tubesock: " + msg);
-      }
-    }
+  private static class StringList {
```

Comment:
I thought for the longest time (up until this very line) that you were using an existing class. How much do we gain from having our own custom implementation, versus just doing it inline in the calling class?

And if you do decide to keep it, a small comment would be nice.

Reply:
We were using a custom [`StringListReader`](https://github.com/firebase/firebase-admin-java/blob/03bc30bc99fbd66068d6c3c3661cd972b6e16de1/src/main/java/com/google/firebase/database/connection/util/StringListReader.java) class for this purpose, which was far more complex than it needed to be. I just created a simpler version of it. Added a class comment to explain its purpose. I do believe the util class makes the core a bit more readable.